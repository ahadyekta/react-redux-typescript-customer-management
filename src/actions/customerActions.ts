// Import redux types
import { ActionCreator, Dispatch } from 'redux';
import { ThunkAction } from 'redux-thunk';


// Import Character Typing
import { ICustomer, ICustomerState } from '../reducers/customerReducer';

// Create Action Constants
export enum CustomerActionTypes {
  CREATE_CUSTOMER = 'CREATE_CUSTOMER',
  EDIT_CUSTOMER = 'EDIT_CUSTOMER',
  DELETE_CUSTOMER = 'DELETE_CUSTOMER',
}

export interface ICustomerCreateAction {
  type: CustomerActionTypes.CREATE_CUSTOMER;
  data: ICustomer;
}

export interface ICustomerEditAction {
    type: CustomerActionTypes.EDIT_CUSTOMER;
    data: ICustomer;
    id: number
}

export interface ICustomerDeleteAction {
    type: CustomerActionTypes.DELETE_CUSTOMER;
    id: number
}

/* 
Combine the action types with a union (we assume there are more)
*/
export type CustomerActions = ICustomerCreateAction | ICustomerEditAction | ICustomerDeleteAction;

/* Create Customer Action
*/
export const createCustomer: ActionCreator<
  ThunkAction<void, ICustomerState, null, ICustomerCreateAction>
> = (data: ICustomer) => {
  return (dispatch: Dispatch) => {
    dispatch({
        data: data,
        type: CustomerActionTypes.CREATE_CUSTOMER,
      });
  };
};


/* Edit Customer Action
*/
export const editCustomer: ActionCreator<
  ThunkAction<void, ICustomerState, null, ICustomerCreateAction>
> = (data: ICustomer, id: Number) => {
  return (dispatch: Dispatch) => {
    dispatch({
        data: data,
        id: id,
        type: CustomerActionTypes.EDIT_CUSTOMER,
      });
  };
};

/* Delete Customer Action
*/
export const deleteCustomer: ActionCreator<
  ThunkAction<void, ICustomerState, null, ICustomerCreateAction>
> = (id: Number) => {
  return (dispatch: Dispatch) => {
    dispatch({
        id: id,
        type: CustomerActionTypes.DELETE_CUSTOMER,
    });
  };
};