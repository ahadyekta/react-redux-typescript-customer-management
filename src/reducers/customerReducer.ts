// Import Reducer type
import { Reducer } from 'redux'
import { CustomerActionTypes, CustomerActions } from '../actions/customerActions'
import customerData from '../data/customerData.json'

// Define the Character type
export interface ICustomer {
    id?: number
    firstName: string
    lastName: string
    DOB: string
}

// Define the Character State
export interface ICustomerState {
    readonly customers: ICustomer[]
}

// Define the initial state
const initialCustomerState: ICustomerState = {
    customers: customerData.customers,
}

export const customerReducer: Reducer<ICustomerState, CustomerActions> = (state = initialCustomerState, action) => {
    switch (action.type) {
        case CustomerActionTypes.CREATE_CUSTOMER: {
            const id =
                state.customers.length > 0
                    ? state.customers[state.customers.length - 1] &&
                      state.customers[state.customers.length - 1].id &&
                      Number(state.customers[state.customers.length - 1].id) + 1
                    : 1 //last customer id
            return {
                ...state,
                customers: [...state.customers, { ...action.data, id }],
            }
        }
        case CustomerActionTypes.DELETE_CUSTOMER: {
            const refinedCustomers = state.customers.filter(({ id }: ICustomer) => id !== action.id)
            return {
                ...state,
                customers: refinedCustomers,
            }
        }
        case CustomerActionTypes.EDIT_CUSTOMER: {
            const refinedCustomers = state.customers.map((customer: ICustomer) => {
                if (customer.id === action.id) {
                    return { ...action.data, id: action.id }
                } else {
                    return { ...customer }
                }
            })
            return {
                ...state,
                customers: refinedCustomers,
            }
        }
        default:
            return state
    }
}
