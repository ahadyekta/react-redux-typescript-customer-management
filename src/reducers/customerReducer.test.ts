import { customerReducer } from './customerReducer'
import { ICustomerCreateAction, ICustomerDeleteAction, ICustomerEditAction } from '../actions/customerActions'

const customers = [
    {
        id: 1,
        firstName: 'Ross',
        lastName: 'Geller',
        DOB: '11/03/1950',
    },
    {
        id: 2,
        firstName: 'Rachel',
        lastName: 'Green',
        DOB: '01/06/1955',
    },
]
describe('Testing customerReducer', () => {
    const state = {
        customers: customers,
    }
    test('When event is CREATE_CUSTOMER, it should add the new customer to list ', () => {
        const action = {
            type: 'CREATE_CUSTOMER',
            data: {
                firstName: 'John',
                lastName: 'Doe',
                DOB: '3/5/1977',
            },
        }
        const result = customerReducer(state, <ICustomerCreateAction>action)
        expect(result.customers).toMatchSnapshot()
    })
    test('When event is DELETE_CUSTOMER, it should delete the item from the list', () => {
        const action = {
            type: 'DELETE_CUSTOMER',
            id: 1,
        }
        const result = customerReducer(state, <ICustomerDeleteAction>action)
        expect(result.customers).toMatchSnapshot()
    })
    test('When event is EDIT_CUSTOMER, it should delete the item from the list', () => {
        const action = {
            type: 'EDIT_CUSTOMER',
            id: 1,
            data: {
                firstName: 'Tom',
                lastName: 'Cruz',
                DOB: '5/7/1967',
            },
        }
        const result = customerReducer(state, <ICustomerEditAction>action)
        expect(result.customers).toMatchSnapshot()
    })
})
