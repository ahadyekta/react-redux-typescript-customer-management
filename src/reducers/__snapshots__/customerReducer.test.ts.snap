// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing customerReducer When event is CREATE_CUSTOMER, it should add the new customer to list  1`] = `
Array [
  Object {
    "DOB": "11/03/1950",
    "firstName": "Ross",
    "id": 1,
    "lastName": "Geller",
  },
  Object {
    "DOB": "01/06/1955",
    "firstName": "Rachel",
    "id": 2,
    "lastName": "Green",
  },
  Object {
    "DOB": "3/5/1977",
    "firstName": "John",
    "id": 3,
    "lastName": "Doe",
  },
]
`;

exports[`Testing customerReducer When event is DELETE_CUSTOMER, it should delete the item from the list 1`] = `
Array [
  Object {
    "DOB": "01/06/1955",
    "firstName": "Rachel",
    "id": 2,
    "lastName": "Green",
  },
]
`;

exports[`Testing customerReducer When event is EDIT_CUSTOMER, it should delete the item from the list 1`] = `
Array [
  Object {
    "DOB": "5/7/1967",
    "firstName": "Tom",
    "id": 1,
    "lastName": "Cruz",
  },
  Object {
    "DOB": "01/06/1955",
    "firstName": "Rachel",
    "id": 2,
    "lastName": "Green",
  },
]
`;
